### PR TITLE
Get Cylc template variables from database for validation.

### LINKED ISSUE #4740: config: template variables specified to play should be loaded
Template variables can be provided to `cylc play`, they are then stored in the DB so they can be used with future restarts.

However, the `cylc config` command is currently unaware of this:

```console
$ cylc install one
$ cylc config
$ cylc play one -s "FOO='bar'"
$ cylc config  # should load FOO='bar' from the workflow DB
```

Note that Cylc-Rose can also be used to specify template variables (at install time), these variables are store in a file which **is loaded** when `cylc config` is invoked.

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by wxtim**
Allowed "revalidation" of cylc workflows where template variables are collected from

Probably closes #4740 

database of already played workflows.
- Added `--revaladidate` as an option to the following scripts (Made async to allow testing of changes):
  - cylc validate
  - cylc view
  - cylc graph
  - cylc config

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Tests are included.
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [ ] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.

### CODE REVIEW

**Comment by wxtim** [comment]
Converted to draft because this change doesn't play nicely with Cylc Rose as it is. :(

File: cylc/flow/option_parsers.py [hunk 1]
```diff
@@ -440,6 +445,13 @@ def add_std_options(self):
                 ),
                 action="store", default=None, dest="templatevars_file")
 
+        if self.revalidate:
+            self.add_std_option(
+                '--revalidate',
+                help="Get template variables from prevous workflow run.",
+                action='store_true', default=False
+            )
```
Comment (oliver-sanders) [inline cylc/flow/option_parsers.py:453]: Were we going to infer this option if the provided workflow is a run directory?

File: cylc/flow/parsec/fileparse.py [hunk 2]
```diff
@@ -251,7 +251,7 @@ def process_plugins(fpath, opts):
             # If you want it to work on sourcedirs you need to get the options
             # to here.
             plugin_result = entry_point.resolve()(
-                srcdir=fpath, opts=opts
+                fpath, opts=opts
```
Comment (oliver-sanders) [inline cylc/flow/parsec/fileparse.py:254]: Note: Double check that cylc-rose is ok with this.

File: cylc/flow/pre_configure/get_old_tvars.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __init__(self, run_dir):
+        self.template_vars = {}
+        self._get_db_template_vars(Path(run_dir))
+
+    def _callback(self, _, row):
+        """Extract key and value and run eval_var on them assigning
+        them to self.template_vars.
+        """
+        self.template_vars[row[0]] = eval_var(row[1])
```
Comment (oliver-sanders) [inline cylc/flow/pre_configure/get_old_tvars.py:41]: Note: check that `eval_var` doesn't require fancy error catching for its other uses in the code.

File: tests/integration/plugins/test_get_old_tvars.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    get_option_parser as graph_gop
+)
+from cylc.flow.scripts.config import (
+    _main as config,
+    get_option_parser as config_gop
+)
+
+
+@pytest.fixture(scope='module')
+def create_workflow(mod_one_conf, mod_flow, mod_scheduler):
```
Comment (oliver-sanders) [inline tests/integration/plugins/test_get_old_tvars.py:43]: Fixtures should generally be nouns unless they return callables, in which case verbs make sense.

E.G. `test_workflow`, `example_workflow`.

File: tests/integration/plugins/test_get_old_tvars.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        (True, 'bar')
+    ]
+)
+async def test_basic(create_workflow, mod_start, revalidate, expect):
+    """It returns a pre-existing configuration if opts.revalidate is True"""
+    opts = create_workflow.opts
+    opts.revalidate = revalidate
+
+    async with mod_start(create_workflow.schd):
+        result = get_old_tvars(create_workflow.schd.workflow_run_dir, opts)
```
Comment (oliver-sanders) [inline tests/integration/plugins/test_get_old_tvars.py:70]: Do we need to call `get_old_tvars` manually here?

If it is done when the Scheduler is started, we can access it as `schd.template_vars`.

**Review by oliver-sanders — COMMENTED** [review]
LGTM

File: cylc/flow/option_parsers.py (hunk 1 above)
Comment (wxtim) [reply cylc/flow/option_parsers.py:453]: Yes

File: cylc/flow/pre_configure/get_old_tvars.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+from cylc.flow.templatevars import eval_var
+from optparse import Values
+from pathlib import Path
+from typing import Union
+
+
+class OldTemplateVars:
+    """Gets template variables stored in workflow database.
+
+    Mirrors the interface used in scheduler.py to get db nfo on restart.
```
Comment (hjoliver) [inline cylc/flow/pre_configure/get_old_tvars.py:29]:
```suggestion
    Mirrors the interface used in scheduler.py to get db info on restart.
```

File: cylc/flow/pre_configure/get_old_tvars.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        self.template_vars[row[0]] = eval_var(row[1])
+
+    def _get_db_template_vars(self, run_dir):
+        dao = CylcWorkflowDAO(str(run_dir / self.DB))
+        dao.select_workflow_template_vars(self._callback)
+
+
+# Entry point:
+def main(srcdir: Union[Path, str], opts: 'Values') -> dict:
+    # We can calculate the source directory here!
```
Comment (hjoliver) [inline cylc/flow/pre_configure/get_old_tvars.py:50]:
```suggestion
```

File: cylc/flow/pre_configure/get_old_tvars.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    """Get options from a previously installed run.
+
+    These options are stored in the database.
+    Calculate the templating language used from the shebang line.
```
Comment (hjoliver) [inline cylc/flow/pre_configure/get_old_tvars.py:54]:
```suggestion
    Get the templating language used from the shebang line.
```

File: cylc/flow/pre_configure/get_old_tvars.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    N.B. The srcdir for this plugin to operate on is a workflow run dir.
+
+    Args:
+        srcdir: The directory of a previously run workflow.
+        opts: Options Object
```
Comment (hjoliver) [inline cylc/flow/pre_configure/get_old_tvars.py:60]:
```suggestion
        opts: Command line options
```
(What it is, is given by the type hint).

File: cylc/flow/scripts/validate.py [hunk 10]
```diff
@@ -80,6 +83,11 @@ def get_option_parser():
         default="live", dest="run_mode",
         choices=['live', 'dummy', 'simulation'])
 
+    parser.add_option(
+        '--revalidate', help="Validate as if for re-install",
```
Comment (hjoliver) [inline cylc/flow/scripts/validate.py:87]:
```suggestion
        '--revalidate', help="Validate as if for re-install",
              i.e. <explain what that means...>
```

**Review by hjoliver — COMMENTED** [review]
Quick skim, looks good.

**Comment by wxtim** [comment]
This will want rebasing on Cylc VIP branch after merge.

File: tests/integration/plugins/test_get_old_tvars.py (hunk 4 above)
Comment (wxtim) [reply tests/integration/plugins/test_get_old_tvars.py:43]: Done

File: tests/integration/plugins/test_get_old_tvars.py (hunk 5 above)
Comment (wxtim) [reply tests/integration/plugins/test_get_old_tvars.py:70]: Nice

File: cylc/flow/scripts/validate.py (hunk 10 above)
Comment (wxtim) [reply cylc/flow/scripts/validate.py]: This is now irrelevent

File: cylc/flow/pre_configure/get_old_tvars.py (hunk 3 above)
Comment (wxtim) [reply cylc/flow/pre_configure/get_old_tvars.py:41]: I'm not sure what you're asking me here.

**Comment by wxtim** [comment]
I'm leaving this in draft format - I prefer the approach I took in https://github.com/cylc/cylc-flow/pull/5214 - and that will probably form the basis for the work on validation for reinstall.

**Pull request closed without merging**